source/common/network/tcp_keepalive_option_impl.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include <netinet/in.h>
+#include <netinet/ip.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+
+#include "envoy/network/listen_socket.h"
+
+#include "common/common/logger.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Network {
+
+// Optional variant of setsockopt(2) optname. The idea here is that if the option is not supported
+// on a platform, we can make this the empty value. This allows us to avoid proliferation of #ifdef.
+typedef absl::optional<int> SocketOptionName;
```

Comment:
This should go in a common header somewhere, instead of copy/pasting it.

I think it might be nice to put all of the socket option preprocessor checking into a single header file. Or better yet, put the declarations in a header, and the definitions in a .cc so all the system headers for the constants don't need to be included in large parts of the codebase. Thoughts on that idea @htuch ?

Reply:
I was considering making this an absl::optional<pair<int, int>> so that it can contain the "level" as well as "optname" value. I think I'd definitely want to do that if I make these a single set of definitions, which I agree might be cleaner.